File: src/js/core/schema/code/immutable.js
define([
  "core/semantics/code/interpreter"
],
function(Interpreter) {
  var ImmutableFactory ;

  var ImmutableList = MakeClass("ImmutableList", null, [],
    function() {
    },
    function(super$) {
      this.initialize = function(items) {
        var self = this; 
        return self.$.items = items;
      };

      this.each = function(block) {
        var self = this; 
        return self.$.items.each(block);
      };
    });

  var ManagedObjectBase = MakeClass("ManagedObjectBase", null, [],
    function() {
    },
    function(super$) {
    });

  var ImmutableFactory = MakeClass("ImmutableFactory", null, [Interpreter.Dispatcher],
    function() {
    },
    function(super$) {
      this.schema = function() { return this.$.schema };

      this.file_path = function() { return this.$.file_path };
      this.set_file_path = function(val) { this.$.file_path  = val };

      this.initialize = function(schema) {
        var self = this; 
        self.$.schema = schema;
        self.$.roots = [];
        self.$.file_path = [];
        return self.setup(self.$.schema);
      };

      this.setup = function(obj) {
        var self = this; 
        return self.dispatch_obj("setup", obj);
      };

      this._get = function(name) {
        var self = this; 
        return self.send(name);
      };

      this.register = function(root) {
        var self = this; 
        return self.$.root = root;
      };

      this.setup_Schema = function(schema) {
        var self = this; 
        return schema.classes().each(function(klass) {
          return self.setup(klass);
        });
      };

      this.setup_Class = function(klass) {
        var self = this; 
        var c;
        c = Class.new(ManagedObjectBase);
        self.dynamic_bind(function() {
          return klass.all_fields().each(function(field) {
            return self.setup(field);
          });
        }, new EnsoHash ({ class: c }));
        self._create_factory_method(klass, c);
        return self._create_initialize_method(klass, c);
      };

      this._create_factory_method = function(klass, c) {
        var self = this; 
        return self.define_singleton_method(function() {
          var args = compute_rest_arguments(arguments, 0);
          return c.new.apply(c, [self].concat(args));
        }, klass.name());
      };

      this._create_initialize_method = function(klass, c) {
        var self = this; 
        var interpreter, key, val;
        interpreter = self;
        return c.define_method(function(factory) {
          var args = compute_rest_arguments(arguments, 1);
          return klass.fields().each_with_index(function(fld, i) {
            if (i >= args.size() && ! fld.optional()) {
              self.raise("Creating immutable object without initializing all required fields");
            }
            if (fld.many()) {
              if (key = Schema.class_key(fld.type())) {
                val = ImmutableSet.new(i < args.size()
                  ? args._get(i)
                  : []);
              } else {
                val = ImmutableList.new(i < args.size()
                  ? args._get(i)
                  : []);
              }
            } else {
              val = i < args.size()
                ? args._get(i)
                : interpreter.default_field_value(fld);
            }
            return self.instance_variable_set(S("@", fld.name()), val);
          });
        }, "initialize");
      };

      this.setup_Field = function(fld) {
        var self = this; 
        if (fld.computed()) {
          return self.setup_computed(self.$.D._get("class"), fld);
        } else {
          return self.define_getter(self.$.D._get("class"), fld);
        }
      };

      this.define_getter = function(c, fld) {
        var self = this; 
        return c.define_method(function() {
          return self.instance_variable_get(S("@", fld.name()));
        }, fld.name().to_sym());
      };

      this.setup_computed = function(c, fld) {
        var self = this; 
        var exp, computed, val;
        exp = fld.computed();
        computed = false;
        return c.define_method(function() {
          if (! computed) {
            val = Impl.eval(exp, new EnsoHash ({ env: Env.ObjEnv.new(self) }));
            computed = true;
          }
          return val;
        }, fld.name());
      };
    });

  ImmutableFactory = {
    new: function(schema) {
      var self = this; 
      return ImmutableFactory.new(schema);
    },

    ImmutableList: ImmutableList,
    ManagedObjectBase: ManagedObjectBase,
    ImmutableFactory: ImmutableFactory,

  };
  return ImmutableFactory;
})

File: src/core/schema/code/immutable.rb
require 'core/semantics/code/interpreter'


module ImmutableFactory

  class ImmutableList
    def initialize(items)
      @items = items
    end

    def each(&block)
      @items.each &block      
    end
  end
  

  class ManagedObjectBase
    class << self
      public :define_method
    end
  end

  def self.new(schema)
    ImmutableFactory.new(schema)
  end

  # every model/graph should have its own factory, which "owns" the nodes of the graph  
  class ImmutableFactory
    attr_reader :schema
    attr_accessor :file_path

    include Interpreter::Dispatcher    
      
    def initialize(schema)
      @schema = schema
      @roots = []
      @file_path = []
      setup(@schema)
    end

    def setup(obj)
      dispatch_obj(:setup, obj)
    end

    # make factory.["Foo"] equivalent to factory.Foo
    def [](name)
      send(name)
    end
    
    def register(root)
      #raise "Creating two roots" if @root
      @root = root
    end
    
    # for every class in the schema, create the factory methods
    def setup_Schema(schema)
      schema.classes.each do |klass|
        setup(klass)
      end
    end

    def setup_Class(klass)
      # all managed objects are subclass of a standard base class
      c = Class.new(ManagedObjectBase)
      dynamic_bind class: c do
        klass.all_fields.each do |field|
          setup(field)
        end
      end
      _create_factory_method(klass, c)
      _create_initialize_method(klass, c)
    end

    # this is a helper method that can be overiden    
    def _create_factory_method(klass, c)
      define_singleton_method(klass.name) do |*args|
        c.new(self, *args)
      end 
    end
    
    # this is a helper method that can be overiden    
    def _create_initialize_method(klass, c)
      # this local variable allows the created initialize method to call back to the
      # data model interpreter
      interpreter = self
      c.define_method(:initialize) do |factory, *args|
        # initialize all the data of the object
        klass.fields.each_with_index do |fld, i|
          if i >= args.size && !fld.optional
            raise "Creating immutable object without initializing all required fields"
          end
          if fld.many then
            # don't use << to insert into the collection, because it should be immutable.
            # create an indexed or non-indexed collection depending on fld.key
            if fld.key
              val = ImmutableSet.new(i < args.size ? args[i] : [])
            else
              val = ImmutableList.new(i < args.size ? args[i] : [])
            end
          else
            val = i < args.size ? args[i] : nil
          end
          instance_variable_set("@#{fld.name}", val)
        end
      end 
    end

    def setup_Field(fld)
      if fld.computed then
        setup_computed(@D[:class], fld)
      else
        define_getter(@D[:class], fld)
      end
    end

    def define_getter(c, fld)
      c.define_method(fld.name.to_sym) do ||
        instance_variable_get("@#{fld.name}")
      end
    end
    
    def setup_computed(c, fld)
      exp = fld.computed
      # compute it once, because it must be constant
      computed = false
      c.define_method(fld.name) do
        if !computed
          val = Impl::eval(exp, env: Env::ObjEnv.new(self))
          computed = true
        end
        val
      end
    end
  end
end
